### ISSUE TITLE
Timeout exception but query was committed retry throws primary key violation

### ISSUE THREAD

**Issue by supremedictator** [aspnet/EntityFramework6#218]
I am using Entity Framework 6.1.2 in an Azure Cloud Service connecting to an Azure SQL Database. I have added both the RetryStrategy and the CommitFailureHandler to our configuration.

When a request is received we are trying to insert a new row into a table using the raw SQL part of Entity Framework.

We first check to see if the row exists for the ID we want to insert.
    
    var rows = dbContext.Database.SqlQuery<T>(
         "SELECT * FROM FactState WHERE ID = @p0 ", ID
    ).ToList();
    return rows != null;`

If it doesn't exist then we insert the row

    dbContext.Database.ExecuteSqlCommand(
        "INSERT INTO FactState ( ID, Value ) VALUES( @p0, @p1 )",
        ID,
        Value
    );

Every once in a while the insert query is throwing a Timeout exception which the retry strategy correctly retries. On the retry however the query throws a Primary Key Violation for each retry until it reaches the limit of 5.

How is this possible, I would have expected a Timeout exception to result in the rows not being added at all, or if the exception occurred during the Commit phase then the commit failure handler would have run and correctly identified the query had succeeded.

I have stepped through the source of Entity Framework in a test program, I can simulate the behavior by manually inserting the row after forcing a timeout with table locks. But I don't understand how it can happen in our production environment.

**Update**

The time out exception has happened multiple times again over the weekend. The exception is happening inside the commit process but this means the commit failure handler is not being run. I suppose its possible the commit failure handler is not attached at the time the commit is running, however I don't understand how given I am using the following code to add the commit failure handler
   
    public class MyConfiguration : DbConfiguration
    {
        public MyConfiguration()
        {
            this.SetExecutionStrategy(
                "System.Data.SqlClient",
                () => new SqlAzureExecutionStrategy());
            
            this.SetTransactionHandler(
                SqlProviderServices.ProviderInvariantName,
                () => new CommitFailureHandler()); 
        }
    }
    

I have attached a stack trace when the error occurred

    Exception: System.Data.SqlClient.SqlException (0x80131904): Timeout expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.  This failure occurred while attempting to connect to the routing destination. The duration spent while attempting to connect to the original server was - [Pre-Login] initialization=1; handshake=18; [Login] initialization=0; authentication=0; [Post-Login] complete=1;   ---> System.ComponentModel.Win32Exception (0x80004005): The wait operation timed out
    at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
    at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
    at System.Data.SqlClient.TdsParserStateObject.ReadSniError(TdsParserStateObject stateObj, UInt32 error)
    at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync()
    at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket()
    at System.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer()
    at System.Data.SqlClient.TdsParserStateObject.TryReadByte(Byte& value)
    at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
    at System.Data.SqlClient.TdsParser.Run(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj)
    at System.Data.SqlClient.TdsParser.TdsExecuteTransactionManagerRequest(Byte[] buffer, TransactionManagerRequestType request, String transactionName, TransactionManagerIsolationLevel isoLevel, Int32 timeout, SqlInternalTransaction transaction, TdsParserStateObject stateObj, Boolean isDelegateControlRequest)
    at System.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransactionYukon(TransactionRequest transactionRequest, String transactionName, IsolationLevel iso, SqlInternalTransaction internalTransaction, Boolean isDelegateControlRequest)
    at System.Data.SqlClient.SqlInternalTransaction.Commit()
    at System.Data.SqlClient.SqlTransaction.Commit()
    at System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch[TTarget,TInterceptionContext](TTarget target, Action`2 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed)
    at System.Data.Entity.Infrastructure.Interception.DbTransactionDispatcher.Commit(DbTransaction transaction, DbInterceptionContext interceptionContext)
    at System.Data.Entity.Core.EntityClient.EntityTransaction.Commit()

**Comment by ajcvickers** [comment]
@AndriySvyryd Can you take a look at this and request more info if needed?

**Comment by AndriySvyryd** [comment]
@supremedictator You can verify that the commit failure handler is configured correctly by using a profiler to monitor you database. It should have a table named `__TransactionHistory`, on every transaction start a row should be inserted there and removed after a successful commit.
Can you check whether the handler is invoked on other operations?
If not could you create a small repro project that exhibits this behavior?

**Comment by supremedictator** [comment]
@AndriySvyryd Thank you for your help
I can confirm there is a __TransactionHistory table and it is getting populated for other operations.
I have used the profiler when the query in question executes and can see it is also inserting a row into the __TransactionHistory table on my local development machine as expected.
And then the row is removed a little later.

I haven't done any profiling on Azure with the Azure SQL Database, I can do this on a test server if required but I am not expecting it to make a difference.

**Comment by AndriySvyryd** [comment]
This looks like a bug/limitation in the commit failure handler. When a commit fails it only tries once to verify it and if this fails it assumes the transaction was rolled back without retrying.

Try using this modified handler:
```
        this.SetTransactionHandler(
            SqlProviderServices.ProviderInvariantName,
            () => new SqlCommitFailureHandler()); 
```

You will also need EF 6.2.0 or use the [SuspendableSqlAzureExecutionStrategy](http://stackoverflow.com/questions/31899382/how-to-use-sqlazureexecutionstrategy-and-nolock/32102380#32102380)

[SqlCommitFailureHandler.cs.txt](https://github.com/aspnet/EntityFramework6/files/859872/SqlCommitFailureHandler.cs.txt)

**Comment by AndriySvyryd** [comment]
@supremedictator Did you have a chance to try the workaround? We are considering to include it in 6.2.0 if it resolves the issue.

**Comment by supremedictator** [comment]
My test program only reproduces the symptoms. So it is difficult to reproduce exactly what our production system is seeing.
The main difference in my test program is I am forcing the timeout exception to happen before the committed method is being called by using TABLOCKX, however when commit doesn't get called then the commit failure handler never gets a chance to verify if the query actually succeeded.

On our production system the timeout must be happening during the commit function (otherwise the rows would never be there to cause a primary key exception on the retry), I haven't found a way to reproduce this behavior in my test environment though.

Using the SqlCommitFailureHandler and the SuspendableSqlAzureExecutionStrategy doesn't resolve the problem in my test environment but I am assuming this is because my test environment is not reproducing the problem the same as our production system as I mentioned above.

If you have any advice on how to force a timeout during the commit method of EntityFramework that would be helpful.
I am unable to deploy the changes to our production system at this stage.

**Comment by AndriySvyryd** [comment]
@supremedictator I've fixed the issue that I believe you were hitting. The fix will be available in EF 6.2.0-beta
